control-plane/connect-inject/controllers/serviceaccount/serviceaccount_controller.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/hashicorp/go-multierror"
+)
+
+type Controller struct {
+	client.Client
+	// ConsulServerConnMgr is the watcher for the Consul server addresses used to create Consul API v2 clients.
+	ConsulServerConnMgr consul.ServerConnectionManager
+	// K8sNamespaceConfig manages allow/deny Kubernetes namespaces.
+	controllers.K8sNamespaceConfig
+	// ConsulNamespaceConfig manages settings related to Consul namespaces.
+	controllers.ConsulNamespaceConfig
+
+	Log logr.Logger
+
+	Scheme *runtime.Scheme
+	context.Context
+}
+
+func (r *Controller) Logger(name types.NamespacedName) logr.Logger {
+	return r.Log.WithValues("request", name)
+}
+
+func (r *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.ServiceAccount{}).
+		Complete(r)
+}
+
+// Reconcile reads the state of a ServiceAccount object for a Kubernetes namespace and reconciles the corresponding
+// Consul WorkloadIdentity.
+func (r *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var errs error
+	var serviceAccount corev1.ServiceAccount
+
+	// Ignore the request if the namespace of the service account is not allowed.
+	if common.ShouldIgnore(req.Namespace, r.DenyK8sNamespacesSet, r.AllowK8sNamespacesSet) {
+		return ctrl.Result{}, nil
+	}
+
+	// Create Consul resource service client for this reconcile.
+	resourceClient, err := consul.NewResourceServiceClient(r.ConsulServerConnMgr)
+	if err != nil {
+		r.Log.Error(err, "failed to create Consul resource client", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
+	}
+
+	// If the ServiceAccount object has been deleted (and we get an IsNotFound error),
+	// we need to deregister that WorkloadIdentity from Consul.
+	err = r.Client.Get(ctx, req.NamespacedName, &serviceAccount)
+	if k8serrors.IsNotFound(err) {
+		err = r.deregisterWorkloadIdentity(ctx, resourceClient, req.Name, r.getConsulNamespace(req.Namespace), r.getConsulPartition())
+		return ctrl.Result{}, err
+	} else if err != nil {
+		r.Log.Error(err, "failed to get ServiceAccount", "name", req.Name, "ns", req.Namespace)
+		return ctrl.Result{}, err
+	}
+	r.Log.Info("retrieved ServiceAccount", "name", req.Name, "ns", req.Namespace)
+
+	// Ensure the WorkloadIdentity exists.
+	if err = r.registerWorkloadIdentity(ctx, resourceClient, &serviceAccount); err != nil {
+		errs = multierror.Append(errs, err)
+	}
+
+	return ctrl.Result{}, errs
+}
+
+func (r *Controller) registerWorkloadIdentity(ctx context.Context, resourceClient pbresource.ResourceServiceClient, a *corev1.ServiceAccount) error {
+	workloadIdentityResource := r.getWorkloadIdentityResource(
+		a.Name, // Consul and Kubernetes service account name will always match
+		r.getConsulNamespace(a.Namespace),
+		r.getConsulPartition(),
+		map[string]string{
+			//TODO verify these keys are correct - update RFC or change to match
+			constants.MetaKeyKubeNS:    a.Namespace,
```

Comment:
Asked about this one [here](https://docs.google.com/document/d/1q01OHEnQ6Ir40y5GeVq53X5iN-5YExHshX_pashLvlQ/edit?disco=AAAA5RR6uhc)

Reply:
In the spirit of simplicity, going to leave this as-is for now (and leave out the RFC meta for the service account controller name, which today always matches k8s like Services). We can always add it later if needed.